aws/cognito_user_pool/variables.tf

```diff
@@ -33,6 +33,38 @@ variable "password_policy" {
   }
 }
 
+variable "lambda_config" {
+  description = "Sets which lambda function to use for various cognito specializations"
+  type = object(
+    {
+      create_auth_challenge          = string
+      custom_message                 = string
+      define_auth_challenge          = string
+      post_authentication            = string
+      post_confirmation              = string
+      pre_authentication             = string
+      pre_sign_up                    = string
+      pre_token_generation           = string
+      user_migration                 = string
+      verify_auth_challenge_response = string
+      kms_key_id                     = string
+    }
+  )
+  default = {
+    create_auth_challenge          = ""
+    custom_message                 = ""
+    define_auth_challenge          = ""
+    post_authentication            = ""
+    post_confirmation              = ""
+    pre_authentication             = ""
+    pre_sign_up                    = ""
+    pre_token_generation           = ""
+    user_migration                 = ""
+    verify_auth_challenge_response = ""
+    kms_key_id                     = ""
+  }
+}
```

Comment:
My preference is to change the type of each argument from `string` to `optional(string)` (https://www.terraform.io/docs/language/expressions/type-constraints.html#experimental-optional-object-type-attributes) and then setting
```
default = {}
```
This does a couple things:
- Makes it possible to pass in
  ```
  lambda_config = {
    pre_sign_up = ...
  }
  ```
  without having to explicitly build the rest of the object with empty strings
- The values of unprovided arguments get set to `null`. In Terraform, when you pass `argument = null`, it essentially ignores the assignment entirely (https://www.terraform.io/docs/language/expressions/types.html#types)

Reply:
This will probably require adding a `terraform.tf` file with the contents:
```
terraform {
  experiments = [module_variable_optional_attrs]
}
```
to enable the `optional` type